### ISSUE TITLE
False positives comparing equality of Maps/Sets with core-js

### ISSUE THREAD

**Issue by mjackson** [ljharb/is-equal#4]
[Using v1.4.0](https://travis-ci.org/mjackson/expect/builds/96360276#L976), I'm seeing a few false positives in Safari 7-8 and IE 9-11. In [the `tmp` branch](https://github.com/mjackson/expect/tree/tmp) of expect I pushed [a commit](https://github.com/mjackson/expect/commit/b6ce6ed68a0584558d6fb478cdf01f439c153fef) that demonstrates the problem. You can see the failed build [here](https://travis-ci.org/mjackson/expect/builds/96360276).

Basically, in these environments, I'm seeing the following:

```js
isEqual({ a: 1 }, { a: 2 }) // true
isEqual({ c: 2 }, { a: 2 }) // true
```

However, both of these values are (correctly) `false` in Chrome 39 and Firefox 32.

**Comment by mjackson** [comment]
I'm not 100% sure, but it seems like this is a dup of #3.

**Comment by ljharb** [comment]
I'm using those test cases, and it's working for me in Safari 7 and 8, and IE 9 - 11 :-/

**Comment by ljharb** [comment]
Is there any possibility that in your tests, you've got something that modifies globals?

**Comment by mjackson** [comment]
> Is there any possibility that in your tests, you've got something that modifies globals?

Yes, there is. I'm [using Babel's polyfill](https://github.com/mjackson/expect/blob/7f79510cd3d49e4931c0dc79b5bef6fdfa1f9253/karma.conf.js#L63) to get Map/Set support in those browsers. That's probably the culprit.

Either:

a) I should just skip Map/Set tests in environments that don't support them (probably easiest) or
b) we should make sure `is-equal` is ok with the Babel polyfill

I'm open to either.

**Comment by mjackson** [comment]
ok, I just pushed a commit that only runs tests for `Map` and `Set` in environments that support them natively (no polyfill). However, it looks like I've still got a problem in IE11 where [this test](https://github.com/mjackson/expect/blob/b53c27fbc7e604db5c2cc2ab54ec5b85ab9211e7/modules/__tests__/toNotEqual-test.js#L17-L23) is [failing](https://travis-ci.org/mjackson/expect/builds/96471618#L1271) with `Set`s.

From what I can tell, it looks like

```js
isEqual(new Set('a'), new Set('b')) === true
```

in IE11. Can you please confirm?

**Comment by ljharb** [comment]
OK, so, core-js (babel's polyfill) definitely does slightly different things than es6-shim's (which is what is-equal is tested with) - they're not spec-compliant. I can definitely try to add tests though to make is-equal work with those.

Regarding IE 11, this is actually a browser bug - `new Set('a').size === 0` is true in IE 11, so both of your sets are empty. There's enough browser bugs with passing an iterable in both `Map` and `Set` that you simply *can not* rely on it working, unless of course you're using the es6-shim. If you change your tests to `var a = new Set(); a.add('a'); var b = new Set(); b.add('b'); isEqual(a, b) === true` will pass as expected.

**Comment by mjackson** [comment]
Thanks for the tip. I've updated my `Set` tests accordingly and everything's green in IE 11. Looks like `is-equal` just needs compat with `core-js`.

**Comment by ljharb** [comment]
@mjackson I've narrowed down the issue to a bug in core-js, and filed the above issue to attempt to resolve it.

**Comment by Poplava** [comment]
I found similar problem in Phantomjs@1.9.19

Problem is that Phantom is not throwing an error while calling ```Map.prototype.forEach``` with object as first arg.

https://github.com/ljharb/is-equal/blob/master/index.js#L68

And the ```tryMapSetEntries``` returns an empty array instead of false.

It looks like PhantomJS issue, but may be it will help you.

**Comment by ljharb** [comment]
@Poplava phantomjs 1.9 has Map and Set? I thought it didn't even have "bind". Are you using `core-js` with it?

**Comment by Poplava** [comment]
@ljharb I didn't use ```core-js```. I've just realized, Map and Set may come from ```babel-polyfill```.

**Comment by ljharb** [comment]
@Poplava `babel-polyfill` is `core-js`.

**Comment by Poplava** [comment]
@ljharb oh, really! I'm sorry... I confused.

**Comment by ljharb** [comment]
@mjackson `v1.4.1` is released - this will ensure both that `core-js` doesn't cause false positives due to `Map`/`Set` support, and also that when https://github.com/zloirock/core-js/issues/144 is fixed, `core-js` `Map`s and `Set`s should just automatically start working with `is-equal`.

**Comment by mjackson** [comment]
:+1: Thanks for the quick turnaround on this, @ljharb
